src/NITTA/LuaFrontend.hs

```diff
@@ -15,460 +11,397 @@ Copyright   : (c) Aleksandr Penskoi, 2019
 License     : BSD3
 Maintainer  : aleksandr.penskoi@gmail.com
 Stability   : experimental
+
+This module analyzes an abstract syntax tree of the Lua language source code, provided by Language.Lua module,
+and stores it into a NITTA's data flow graph.
+
+Supported Lua costructions are:
+ - simple math operators (addition, subtraction, multiplication and division);
+ - variable assignments;
+ - bitwise left and right shifts;
+ - recursive calls.
 -}
 module NITTA.LuaFrontend (
     lua2functions,
     FrontendResult (..),
     TraceVar (..),
+
+    -- * Internal
+    LuaAlgBuilder (..),
+    LuaStatement (..),
+    LuaValueInstance (..),
+    findStartupFunction,
+    getLuaBlockFromSources,
+    processStatement,
 ) where
 
-import Control.Monad.Identity
 import Control.Monad.State
-import Data.Bifunctor
 import qualified Data.HashMap.Strict as HM
-import Data.List (find, group, sort)
+import Data.Hashable
 import qualified Data.Map.Strict as M
 import Data.Maybe
 import Data.String
 import Data.String.ToString
-import qualified Data.String.Utils as S
-import Data.Text (Text, pack, unpack)
 import qualified Data.Text as T
 import Language.Lua
 import NITTA.Intermediate.DataFlow
 import qualified NITTA.Intermediate.Functions as F
-import NITTA.Utils (modify'_)
+import NITTA.Utils.Base
 import Text.Printf
 
+getUniqueLuaVariableName LuaValueInstance{luaValueInstanceName, luaValueInstanceIsConstant = True} luaValueAccessCount = "!" <> luaValueInstanceName <> "#" <> showText luaValueAccessCount
+getUniqueLuaVariableName LuaValueInstance{luaValueInstanceName, luaValueInstanceAssignCount} luaValueAccessCount
+    | T.head luaValueInstanceName == '_' = luaValueInstanceName
+    | otherwise = luaValueInstanceName <> "^" <> showText luaValueInstanceAssignCount <> "#" <> showText luaValueAccessCount
+
+data LuaStatement x = LuaStatement
+    { fIn :: [T.Text]
+    , fOut :: [LuaValueInstance]
+    , fName :: T.Text
+    , fValues :: [x]
+    , fInt :: [Int]
+    }
+    deriving (Show, Eq)
+
+-- | Stores information about a particular version of a variable. The version of a variable changes after assigning a new value to it.
+data LuaValueInstance = LuaValueInstance
+    { luaValueInstanceName :: T.Text
+    , luaValueInstanceAssignCount :: Int
+    , luaValueInstanceIsConstant :: Bool
+    }
+    deriving (Show, Eq)
+
 data FrontendResult v x = FrontendResult
     { frDataFlow :: DataFlowGraph v x
     , frTrace :: [TraceVar]
     , frPrettyLog :: [HM.HashMap v x] -> [HM.HashMap String String]
```

Comment:
Maybe, here we should have: `HM.HashMap String String`. I think it will be better if you create an issue for newcomers to do this.

Reply:
https://github.com/ryukzak/nitta/issues/186

Reply:
Looks like I make some mistake in the comment. I will fix issue.